File: 1365-how-many-numbers-are-smaller-than-the-current-number/1365-how-many-numbers-are-smaller-than-the-current-number.cpp
class Solution {
public:
    vector<int> smallerNumbersThanCurrent(vector<int>& nums) {
        vector<int> sorted = nums;
        sort(sorted.begin(), sorted.end());
        int n = sorted.size();
        unordered_map<int, int> mp;

        for (int i=0; i<n; ++i){
            if (mp.find(sorted[i]) == mp.end()){    //so if the number is already present in the map, it will just skip it.
                mp[sorted[i]] = i;  //index of the current number gives the number of numbers smaller that itself.
            }
        }
        vector<int> res;
        for (int num : nums){
            res.push_back(mp[num]);
        }
        return res;
    }
};

File: 1365-how-many-numbers-are-smaller-than-the-current-number/1365-how-many-numbers-are-smaller-than-the-current-number.cs
public class Solution {
    public int[] SmallerNumbersThanCurrent(int[] nums) {
        int[] sorted = (int[])nums.Clone(); //clone arraya, so we don't change original array
        Array.Sort(sorted); //sort cloned array
        Dictionary<int, int> map = new Dictionary<int, int>();

        //For each number in nums, store its index in the Dic. (which equals how many numbers are smaller than that one)
        for(int i=0; i < sorted.Length; i++){
            if (!map.ContainsKey(sorted[i])){
                map[sorted[i]] = i; //map[sorted[i]] -> key | i -> value
            }
        }

        //Build the res using the indexes in the Dic.
        int[] res = new int[nums.Length];
        for (int i=0; i < nums.Length; i++){
            res[i] = map[nums[i]];
        }

        return res;
    }
}